File: ruby/05_from_push_to_pull/lib/pushing_balances_calculator.rb
require_relative 'values_of_month'

class PushingBalancesCalculator
  attr_reader :transactions
  private :transactions
  def initialize(transactions)
    @transactions = transactions
  end

  def fill_data(balances_of_month_list)
    # values_of_month = ValuesOfMonth.new
    balance = 0

    balances_of_month_list.each do |balances_of_month|
      date_of_month = balances_of_month.date
      transactions_of_month = transactions_of_month(date_of_month)

      values_of_month = calculate_values_for_month(balance, date_of_month, transactions_of_month)
      balance = values_of_month.balance

      balances_of_month.balance = values_of_month.balance
      balances_of_month.average_balance = values_of_month.average_balance
    end
  end

  private

  def calculate_values_for_month(preceding_balance, date_of_month, transactions_of_month)
    values_of_month = ValuesOfMonth.new

    values_of_month.calculate_values(date_of_month, transactions_of_month, preceding_balance)
    values_of_month
  end

  def calculate_proportional_balance(day_of_latest_balance, balance, day, days_in_month)
    counting_days = day - day_of_latest_balance
    return 0 if counting_days == 0
    rate = counting_days.to_f / days_in_month.to_f
    balance * rate
  end

  def transactions_of_month(date)
    transactions.select { |transaction|
      are_same_month_and_year(date, transaction.date)
    }
  end

  def are_same_month_and_year(date, date_of_transaction)
    date_of_transaction.month == date.month &&
      date_of_transaction.year == date.year
  end
end

File: java/03FromPushToPull/src/push/PushingBalancesCalculator.java
package push;

import java.util.ArrayList;
import java.util.List;

import org.joda.time.LocalDate;

import common.BalancesOfMonth_API;
import common.BalancesOfMonthCalculator_API;
import common.Transaction_API;

public class PushingBalancesCalculator implements BalancesOfMonthCalculator_API
{

	private final List<Transaction_API> transactions;

	public PushingBalancesCalculator(List<Transaction_API> transactions)
	{
		super();
		this.transactions = transactions;
	}

	@Override
	public void fillData(List<BalancesOfMonth_API> balancesOfMonthList)
	{
		ValuesOfMonth valuesOfMonth = new ValuesOfMonth();
		int balance = 0;

		for (BalancesOfMonth_API balancesOfMonth : balancesOfMonthList)
		{
			LocalDate dateOfMonth = balancesOfMonth.getDate();
			List<Transaction_API> transactionsOfMonth = transactionsOfMonth(dateOfMonth);

			balance = calculateValuesForMonth(valuesOfMonth, balance, dateOfMonth, transactionsOfMonth);

			balancesOfMonth.setBalance(valuesOfMonth.getBalance());
			balancesOfMonth.setAverageBalance(valuesOfMonth.getAverageBalance());
		}
	}

	private int calculateValuesForMonth(ValuesOfMonth valuesOfMonth, int precedingBalance, LocalDate dateOfMonth, List<Transaction_API> transactionsOfMonth)
	{
		int balance = precedingBalance;
		int ultimo = dateOfMonth.getDayOfMonth();

		double averageBalance = 0;
		int dayOfLatestBalance = 1;
		for (Transaction_API transaction : transactionsOfMonth)
		{
			int day = transaction.getDate().getDayOfMonth();
			averageBalance += calculateProportionalBalance(dayOfLatestBalance, balance, day, ultimo);
			balance += transaction.getAmount();
			dayOfLatestBalance = day;
		}

		averageBalance += calculateProportionalBalance(dayOfLatestBalance, balance, ultimo + 1, ultimo);

		valuesOfMonth.setBalanceAndAverage(balance, averageBalance);
		return balance;
	}

	private double calculateProportionalBalance(int dayOfLatestBalance, int balance, int day, int daysInMonth)
	{
		int countingDays = day - dayOfLatestBalance;
		if (countingDays == 0)
		{
			return 0;
		}
		double rate = (double) countingDays / daysInMonth;
		return (balance * rate);
	}

	private List<Transaction_API> transactionsOfMonth(LocalDate date)
	{
		List<Transaction_API> results = new ArrayList<Transaction_API>();
		for (Transaction_API transaction : transactions)
		{
			LocalDate dateOfTransaction = transaction.getDate();
			if (areSameMonthAndYear(date, dateOfTransaction))
			{
				results.add(transaction);
			}
		}
		return results;
	}

	private boolean areSameMonthAndYear(LocalDate date, LocalDate dateOfTransaction)
	{
		return dateOfTransaction.getMonthOfYear() == date.getMonthOfYear() && dateOfTransaction.getYear() == date.getYear();
	}

}
